### ISSUE TITLE
add-attachment throw Runtime Error

### ISSUE THREAD

**Issue by julig** [pycontribs/jira#1004]
python-version 3.8 

Code-Snippet:
with open(filename, 'rb') as f:
        jira_handler.add_attachment(issue=issue], attachment=f)
throw Runtime-Error:
client.py", line 787, in add_attachment
    url, data=m, headers=CaseInsensitiveDict({'content-type': m.content_type, 'X-Atlassian-Token': 'nocheck'}), retry_data=file_stream)

jira\utils\__init__.py", line 41, in __init__
    for key, value in super(CaseInsensitiveDict, self).items():
RuntimeError: dictionary keys changed during iteration

**Comment by EdgeSync** [comment]
I also started getting this error a number of days ago - but this is on Ubuntu 18.04, using Python 3.6.9.
Not sure if it is related.

`2021-03-10 17:15:14,721+0000 https-jsse-nio-8443-exec-4 ERROR ***REDACTED*** 1035x776013x1 1n0ssss 16.208.80.13 /rest/api/2/issue/GIA-3244/attachments [c.a.j.r.v2.issue.IssueAttachmentsResource] Error saving attachment
java.lang.NullPointerException
        at com.atlassian.jira.issue.util.DefaultIssueUpdater.storeModifiedFields(DefaultIssueUpdater.java:78)
        at com.atlassian.jira.issue.util.DefaultIssueUpdater.doUpdate(DefaultIssueUpdater.java:69)
        at com.atlassian.jira.issue.util.DefaultIssueUpdater.doUpdate(DefaultIssueUpdater.java:48)
        at sun.reflect.GeneratedMethodAccessor2127.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
        at com.sun.proxy.$Proxy461.doUpdate(Unknown Source)
        ... 2 filtered
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:343)
        at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
        at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:136)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:124)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
        at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
        at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
        at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:136)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:124)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
        at com.sun.proxy.$Proxy2866.doUpdate(Unknown Source)
        at com.atlassian.jira.rest.v2.issue.IssueAttachmentsResource.addAttachment(IssueAttachmentsResource.java:128)
        ... 2 filtered
        at java.lang.reflect.Method.invoke(Method.java:498)
        ... 5 filtered
        at com.atlassian.jira.rest.exception.ExceptionInterceptor.intercept(ExceptionInterceptor.java:55)
        ... 1 filtered
        at com.atlassian.jira.rest.v2.issue.scope.RequestScopeInterceptor.intercept(RequestScopeInterceptor.java:39)
        ... 14 filtered
        at com.atlassian.plugins.rest.module.RestDelegatingServletFilter$JerseyOsgiServletContainer.doFilter(RestDelegatingServletFilter.java:159)
        ... 1 filtered
        at com.atlassian.plugins.rest.module.RestDelegatingServletFilter.doFilter(RestDelegatingServletFilter.java:69)
        ... 36 filtered
        at com.atlassian.jira.plugin.mobile.web.filter.MobileAppRequestFilter.doFilter(MobileAppRequestFilter.java:37)
        ... 4 filtered
        at com.atlassian.jira.plugin.mobile.login.MobileLoginSuccessFilter.doFilter(MobileLoginSuccessFilter.java:54)
        ... 3 filtered
        at com.atlassian.diagnostics.internal.platform.monitor.http.HttpRequestMonitoringFilter.doFilter(HttpRequestMonitoringFilter.java:55)
        ... 8 filtered
        at com.atlassian.web.servlet.plugin.request.RedirectInterceptingFilter.doFilter(RedirectInterceptingFilter.java:21)
        ... 57 filtered
        at com.atlassian.jira.security.JiraSecurityFilter.lambda$doFilter$0(JiraSecurityFilter.java:66)
        ... 1 filtered
        at com.atlassian.jira.security.JiraSecurityFilter.doFilter(JiraSecurityFilter.java:64)
        ... 16 filtered
        at com.atlassian.plugins.rest.module.servlet.RestSeraphFilter.doFilter(RestSeraphFilter.java:37)
        ... 19 filtered
        at com.atlassian.jira.servermetrics.CorrelationIdPopulatorFilter.doFilter(CorrelationIdPopulatorFilter.java:30)
        ... 10 filtered
        at com.atlassian.web.servlet.plugin.request.RedirectInterceptingFilter.doFilter(RedirectInterceptingFilter.java:21)
        ... 4 filtered
        at com.atlassian.web.servlet.plugin.LocationCleanerFilter.doFilter(LocationCleanerFilter.java:36)
        ... 26 filtered
        at com.atlassian.jira.servermetrics.MetricsCollectorFilter.doFilter(MetricsCollectorFilter.java:25)
        ... 25 filtered
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)`

**Comment by yunhao-qing** [comment]
Getting same error with 
```
for key, value in super(CaseInsensitiveDict, self).items():
RuntimeError: dictionary keys changed during iteration
```

**Comment by pertoft** [comment]
+1

**Comment by mtripoli** [comment]
+1
This issue is a duplicate of #890 which is fixed in master (3.0). Can this fix be integrated to 2.0?

**Comment by EdgeSync** [comment]
#890 is a different issue to what i've been experiencing. I have found that explicitly setting the filename in my request has solved the issue for me. I wonder did something change on the Linux side, as it seems different versions of Python are affected.

`def add_my_attachment(issue_key, file_name):
    jira.add_attachment(issue_key, file_name, filename=file_name.split('/')[-1]`

**Comment by BHappach** [comment]
> #890 is a different issue to what i've been experiencing. I have found that explicitly setting the filename in my request has solved the issue for me. I wonder did something change on the Linux side, as it seems different versions of Python are affected.
> 
> ```
> jira.add_attachment(issue_key, file_name, filename=file_name.split('/')[-1]
> ```

tried to do so - did not work 4 me

greets Bernd

**Comment by mtripoli** [comment]
I've resorted to monkey-patching `jira.client.CaseInsensitiveDict` by copying the [class definition](https://github.com/pycontribs/jira/blob/master/jira/utils/__init__.py#L8) found in the master branch into my project's utilites module: 

```
import jira.client
from my_local_utils import CaseInsensitiveDict
jira.client.CaseInsensitiveDict = CaseInsensitiveDict
```

Doing this eliminates the `RuntimeError: dictionary keys changed during iteration` exception reported in the description of this issue.

**Comment by rena2damas** [comment]
@mtripoli funny, I got to the exact same solution when answering a StackOverflow [answer](https://stackoverflow.com/a/67332687/9187350).

**Comment by studioj** [comment]
@rena2damas @mtripoli @BHappach @EdgeSync @pertoft @yunhao-qing @julig now version 3.0.1 is released is this still happening, as @mtripoli suggested it could've been fixed in 3.0 ?

**Comment by rena2damas** [comment]
> @rena2damas @mtripoli @BHappach @EdgeSync @pertoft @yunhao-qing @julig now version 3.0.1 is released is this still happening, as @mtripoli suggested it could've been fixed in 3.0 ?

Code looks good, and the latest version is working for me. I can safely remove my workaround. Thanks ! 👍
